DS-Homebrew/nds-bootstrap
# Add AP patches and its loading mechanism

#### What's changed?

As of commit 393f838004da7 from TWiLightMenu. Fully implements and closes #1440.

The code is pretty much a direct copy from TWiLight, with the fcopy BSS a little smaller (we don't need to alloc 32KB of memory for copying, the files are on average less than 512 bytes. So only 512 bytes are alloc'd for now.)

The packaging of the apfixes is removed, and built directly into NitroFS instead. Not much is lost here, as the only reason we were using that packing was solely because small file sizes meant slower extraction for Universal Updater and the like.

Ideally, we'd place the config function in `loadFromSD`. But it keeps crashing there, but it doesn't crash in `load_conf`. So just keep it there for now.

TODO: 
- [x] Implement the AP patching
- [x] Implement AP patch overriding
  - [x] by gameTID:headerCRC
  - [x] by filename
- [ ] Remove the code from TWiLight

For people who want to test:
1. Nuke all AP fix packs and override files from the `TWiLightMenu/extras` folder
2. Build this branch (or DL from GitHub Actions I guess) and place build on SD card appropriately
3. Turn on debug. AP fix path will be shown in the debug screen.
4. Run game with known AP fix and see if it behaves (I am testing a clean Black 2 dump.)

#### Where have you tested it?

Nintendo DSi
needs more testing, such as overriding and/or flashcard with SCFG access etc.

***

#### Pull Request status
- [x] This PR has been tested using the latest version of devkitARM and libnds.
